connect/src/com/telenor/connect/headerenrichment/MobileDataFetcher.java

```diff
@@ -0,0 +1,140 @@
+                connection.setInstanceFollowRedirects(false);
+                connection.connect();
+                int responseCode = connection.getResponseCode();
+                attempts += 1;
+                if (responseCode != HTTP_SEE_OTHER
+                        && responseCode != HTTP_MOVED_TEMP
+                        && responseCode != HTTP_MOVED_PERM) {
+                    // Rely on the WebView to close the input stream when finished fetching data
+                    return new WebResourceResponse(
+                            connection.getContentType(),
+                            connection.getContentEncoding(),
+                            connection.getResponseCode(),
+                            connection.getResponseMessage(),
+                            getHeadersAsCommaJoinedList(connection.getHeaderFields()),
+                            connection.getInputStream());
+                }
+                newUrl = connection.getHeaderField("Location");
+                // Close the input stream, but do not disconnect the connection as its socket might
+                // be reused during the next request.
+                connection.getInputStream().close();
+            } catch (final IOException e) {
+                return null;
+            }
+            if (allowedToToggleNetworkToUse) {
+                interfaceToUse = shouldFetchThroughCellular(newUrl)
+                        ? HeLogic.getCellularNetwork()
+                        : HeLogic.getDefaultNetwork();
+            }
+        } while (attempts <= MAX_REDIRECTS_TO_FOLLOW_FOR_HE);
+        return null;
+    }
+
+    private static Map<String, String> getHeadersAsCommaJoinedList(Map<String, List<String>> headerFields) {
+        Map<String, String> result = new HashMap<>();
+        for (String header : headerFields.keySet()) {
+            List<String> strings = headerFields.get(header);
+            if (strings == null || strings.isEmpty()) {
+                continue;
+            }
+            String commaJoinedList = TextUtils.join(",", strings);
+            result.put(header, commaJoinedList);
+        }
+        return result;
+    }
+
+    public static boolean shouldFetchThroughCellular(String url) {
+        WellKnownAPI.WellKnownConfig wellKnownConfig = ConnectSdk.getWellKnownConfig();
+        if (wellKnownConfig == null ||
+                (wellKnownConfig.getNetworkAuthenticationTargetIps().isEmpty()
+                        && wellKnownConfig.getNetworkAuthenticationTargetUrls().isEmpty())) {
+            return false;
+        }
+        if (!wellKnownConfig.getNetworkAuthenticationTargetUrls().isEmpty()) {
+            for (String urlPrefix : wellKnownConfig.getNetworkAuthenticationTargetUrls()) {
+                if (url.contains(urlPrefix)) {
+                    return true;
+                }
+            }
+            return false;
+        } else {
```

Comment:
Unneccesary else as all above returns anyways.

Reply:
You are correct - points to Gryffindor ;). (btw; it was moved code) 65b0b8a